CODINGSTYLE.md

```diff
@@ -0,0 +1,443 @@
+# Coding Style Guide
+
+The Scala.js project has a strict policy regarding coding style.
+This is one of the cornerstones that has allowed Scala.js to maintain a clean, consistent and maintainable codebase.
+
+This document tries to document the style in use as much as possible to make it easier for everyone to contribute.
+
+A *few* of these rules are checked automatically using Scalastyle, but most of them are too complex to teach to an automated tool.
+
+Our coding style is driven and "maintained" by [@sjrd](https://github.com/sjrd), the original author of Scala.js.
+Please follow his directions if they are in conflict with the present document.
+Feel free to point out the conflict, though: we are always looking to improve this document.
+
+
+## General style (whitespaces, braces, etc.)
+
+### Tabs, new lines, and eofs
+
+* Files must not contain tabs
+* New lines are UNIX style (\n)
+* A final new line must be present at the end of file
+* There must not be whitespace characters at the end of a line
+
+### Line length
+
+Lines should be limited at 80 characters.
+In some cases, if breaking a line makes it significantly less readable, it can go up to 120 characters.
+
+Rationale: when reviewing on GitHub, only 120 characters are visible; when reviewing on a mobile phone, only 80 characters are visible.
+And we do review on mobile phone quite a lot.
+
+#### Where to break a line
+
+A line can be broken after either a `,` or a `(`, or possibly after a binary operator in a long expression.
+
+### Indentation
+
+In general, indentation is 2 spaces, except continuation lines, which are indented 4 spaces.
+A continuation line is a line appearing because we broke something that should have been on one line for line length reasons.
+Typically, this means inside parentheses (formal or actual parameters of methods and constructors), and a long `extends` clause.
+
+Note that breaking a line right after the `=` sign of an initialization or assignment is *not* considered a continuation line.
+
+### Blank lines
+
+* Never put two blank lines in a row
+* (Almost) always put a blank line between two declarations in a class
+* Always put blank lines around a `case` whose body spans several lines
+* Insert blank lines at will between logical blocks of statements in a method
+
+The blank line between two consecutive declarations in a class can sometimes be omitted, if the declarations are single-line (which also means ScalaDocless) and strongly related.
+This happens pretty rarely (mostly a series of private fields).
+The rule of thumb is to always put a blank line.
+
+### Braces
+
+Braces should not be used for one-liner methods.
+Similarly, you should not use braces for the body of a `while`/`if`/`for` that is a one-liner.
+
```

Comment:
Ah, yep, these probably deserve their own section. FTR:
```scala
val myVal = 
  this.sign == that.sign && 
  this.numberLength == that.numberLength && 
  this.equalsArrays(that.digits) && 
  this.equalsArrays(whatever.digits)  // 2 or 4 ?
```
is allowed with 2 spaces, and with braces around it.
```scala
if(this.sign == that.sign && this.numberLength == that.numberLength && 
   this.equalsArrays(that.digits) && this.equalsArrays(whatever.digits)){ // 2 or 4?
```
is allowed, with a space after `if`, and 4 spaces.
More than 2 lines should be avoided, and the condition extracted in a `def` before the `if` instead, with the style of the first snippet.
```scala
val cond1 = this.sign == that.sign && this.numberLength == that.numberLength
val cond2 = this.equalsArrays(that.digits) && this.equalsArrays(whatever.digits)
val myVal = cond1 && cond2
```
is allowed if `cond1` and `cond2` are meaningful and can be given a meaningful name, in which case it is preferred. Otherwise, the style of the first snippet is preferred.